### DISCUSSION TITLE
Porting to other platforms

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#11] [General]
Greetings,

In order to simplify porting of the system to other platforms, I ripped out a lot of the UI code, which was very MacOS/ccl specific.

In **kons.asd** make sure **minimal-ui** is included and **ui** is commented out.

All the device/graphics code should be in **minimal-ui.lisp** and **opengl.lisp**

Maybe to avoid overlap of effort, please reply in this thread if you are going to try a port and let us know what your target system is. Alternately, maybe make a ticket in the project. I don't know yet if you guys can make tickets or if I have to do that.

Good luck! :)

Kaveh

**Comment by omarbassam88** [comment]
I'd be more than happy to do the port to sbcl on (I'm on Linux). However, I believe the issues lie in the following files.
`ccl64.lisp` may this should be replaced by a `platform` or `implementation` file where loads different library depending on the implementation.
Also, in `package.lisp` you are using `:ccl` may we should have an abstraction layer for the implementations as well.

**Reply by kaveh808** [reply]
Feel free to make any changes you deem necessary.

**Reply by mtreis86** [reply]
I started listing things that will need to be changed for SBCL to work, I'll open a new issue for it as they're likely independent of the platform

**Comment by omarbassam88** [comment]
Also, I was wondering if there will be plans to support Vulkan (besides or instead of Open GL)?

**Reply by kaveh808** [reply]
Yes, we will investigate other APIs as well, including a WebGL interface. I think for now the simplest route to porting would be to keep OpenGL.

**Comment by omarbassam88** [comment]
I belive we have two main issues we need to discuss:
1- the list of files (functions) that need to address different implementations.
a reference to how this should be done I guess is this chapter from "Practical Common Lisp": 
- [15. Practical: A Portable Pathname Library](https://gigamonkeys.com/book/practical-a-portable-pathname-library.html)

2- the use of a different UI Framework other than "COCOA" that work on most Platforms. Other Alternatives that I found:

- [imgui](https://github.com/serialhex/cl-imgui)
- [QTools](https://github.com/Shinmera/qtools/https://github.com/Shinmera/qtools/)
- [commonqt](https://github.com/commonqt/commonqt)
- [cl-cffi-gtk](https://github.com/crategus/cl-cffi-gtk)

**Reply by kaiwulf** [reply]
I'm familiar with gtk using C/C++. Using OpenGL with gtk isn't so straightforward. I'm not familiar with imgui or Qt. GLFW is a commonly used windowing library used with OpenGL on Linux. There's [cl-glfw3](https://github.com/AlexCharlton/cl-glfw3). What I've looked at so far, it seems like a suitable option for the Linux port.

**Reply by kaveh808** [reply]
The GUI should be its own module, maybe even with its own namespace. I say this because if we succeed in developing a good cross-platform GUI system for CL, we will be doing the community a huge favor. There is nothing in open source that does this right now.

My inclination is to have a go with glfw. It seems simpler than imgui, and is written in C as opposed to C++.

Over time, we could implement more of the UI features in CL, working our way towards a "CL all the way down" system.

**Comment by awolven** [comment]
imgui and glfw are apples and oranges

glfw is a cross platform windowing/event library and imgui is a widget/text
library

**Reply by kaveh808** [reply]
Good point. My bad, I had a quick glance at the glfw page.

Let's get the minimal-ui to work with glfw as a first step. Then we can explore developing our own widgets.

**Comment by awolven** [comment]
BTW, |3b| pointed out to me that there is a pure common lisp glfw-like
library out there.

**Reply by JMC-design** [reply]
Did he mention which?  I kinda find that hard to believe.  I have not been able to get a gl context from lisp, though it hasn't been my focus since that comes after the windowing basics.

**Comment by JMC-design** [comment]
I briefly looked at GLFW and there might be a problem depending on your use case. I don't usually do ffi, but was thinking it might be nice to have a cross-platform backend to my surface https://github.com/JMC-design/surface protocol.
But GLFW gives you a window and a context tied together.  This might mean you can only do your gui in 3d, which is fine for a 3d program, maybe not so fine for a general purpose gui library.
I'll see if I can use my other stuff to draw on the windows provided by glfw if I can get glfw to actually give me a window as it keeps erroring.

**Comment by awolven** [comment]
|3b| said it was glop or glop2.

**Reply by JMC-design** [reply]
ah but that's not pure CL, that's a bunch of ffi.

**Comment by awolven** [comment]
Since you're aiming for a Blender caliber application, I'm of the stance
that a non-graphics-langauge-based-UI is insufficient for the task.  By a
non-graphics-langauge-based-ui, I mean things like Qt, Win32, Cocoa, CLX,
etc.  The developers of Mirai came up with their own UI, called "Human
Interface", or HI, for short written directly in OpenGL, and inherently
cross-platform.  The choice of Graphics Language itself is not necessarily
relevant toward building a native-graphics-langauge-based-ui.  I would like
to point out Alloy as an example by Nicholas Hafner (I think), and a lot of
other libraries by Nicholas Halfner (aka Shinmera) who has put a ton of
thought into cutting edge common lisp graphics.  I think the "state of the
art" in CL graphics should be researched before building code on top of
less capable toolkits.
